hw/ip/otbn/util/static_check.py

```diff
@@ -0,0 +1,150 @@
+    '''
+    if insn.errs is None:
+        return True
+    for err in insn.errs:
+        if 'CALL_STACK' not in err:
+            continue
+        if err == 'A `CALL_STACK` error from using `x1` as `grd` when the call stack is full and `grs1` is not `x1`.':
+            if operands['grd'] == 1 and not operands['grs1'] == 1:
+                return False
+        elif err == 'A `CALL_STACK` error from using `x1` as `grd` when the call stack is full and neither `grs1` nor `grs2` is `x1`.':
+            if operands['grd'] == 1:
+                if not (operands['grs1'] == 1 or operands['grs2'] == 1):
+                    return False
+        elif err == 'A `CALL_STACK` error from using `x1` as `grs` or `grd` when the call stack is empty.':
+            if operands['grs'] == 1 or operands['grd'] == 1:
+                return False
+        elif err == 'A `CALL_STACK` error from using `x1` as `grd` when the call stack is full.':
+            if operands['grd'] == 1:
+                return False
+        elif err == 'A `CALL_STACK` error from using `x1` as `grs1` or `grs2` when the call stack is empty.':
+            if operands['grs1'] == 1 or operands['grs2'] == 1:
+                return False
+        elif err == 'A `CALL_STACK` error from using `x1` as `grs1` or `grd` when the call stack is empty.':
+            if operands['grs1'] == 1 or operands['grd'] == 1:
+                return False
+        elif err == 'A `CALL_STACK` error from using `x1` as `grs1` when the call stack is empty.':
+            if operands['grs1'] == 1:
+                return False
+        else:
+            raise RuntimeError(
+                'Unrecognized CALL_STACK error condition: {}'.format(err))
+    return True
+
+
+def check_call_stack(program: OTBNProgram) -> Tuple[bool, str]:
+    '''Check that the special register x1 is used safely.
+
+    If x1 is used for purposes unrelated to the call stack, it can trigger a
+    CALL_STACK error. This check rules out such errors for any instructions
+    other than JAL or JALR.
+    '''
+    errs_allowed = ['jal', 'jalr']
+    imem_start = program._get_pc_at_symbol('_imem_start')
+    imem_end = program._get_pc_at_symbol('_imem_end')
+    for pc in range(imem_start, imem_end, 4):
+        insn = program._get_insn(pc)
+        if insn.mnemonic in errs_allowed:
+            continue
+        operands = program._get_operands(pc)
+        if not _check_call_stack_insn(insn, operands):
+            return (False, 'check_call_stack: FAIL at PC {:#x}: {} {}'.format(
+                pc, insn.mnemonic, operands))
+    return (True, 'check_call_stack: PASS')
+
+
+def main() -> int:
+    parser = argparse.ArgumentParser()
+    parser.add_argument('elf', help=('The .elf file to check.'))
+    args = parser.parse_args()
+    program = decode_elf(args.elf)
+    ok, msg = check_call_stack(program)
+    print(msg)
+    if not ok:
+        return 1
+    return 0
```

Comment:
Maybe `return 1 if not ok else 0` is shorter but still explicit on the bool -> return code conversion.